src/core/listener.c

```diff
@@ -918,18 +1040,58 @@ _IRQL_requires_max_(DISPATCH_LEVEL)
 void
 QuicListenerHandleDosModeStateChange(
     _In_ QUIC_LISTENER* Listener,
-    _In_ BOOLEAN DosModeEnabled
+    _In_ BOOLEAN DosModeEnabled,
+    _In_ BOOLEAN OnWorker
     )
 {
     if (Listener->DosModeEventsEnabled) {
-        QUIC_LISTENER_EVENT Event;
-        Event.Type = QUIC_LISTENER_EVENT_DOS_MODE_CHANGED;
-        Event.DOS_MODE_CHANGED.DosModeEnabled = DosModeEnabled;
+        if (!Listener->Partitioned || OnWorker) {
+            QUIC_LISTENER_EVENT Event;
+            Event.Type = QUIC_LISTENER_EVENT_DOS_MODE_CHANGED;
+            Event.DOS_MODE_CHANGED.DosModeEnabled = DosModeEnabled;
 
-        QuicListenerAttachSilo(Listener);
+            QuicListenerAttachSilo(Listener);
 
-        (void)QuicListenerIndicateDispatchEvent(Listener, &Event);
+            (void)QuicListenerIndicateDispatchEvent(Listener, &Event);
 
-        QuicListenerDetachSilo();
+            QuicListenerDetachSilo();
+        } else {
+            //
+            // Best effort mode synchronization: the non-partitioned case is
+            // also racy.
+            //
+            Listener->DosModeEnabled = DosModeEnabled;
+            if (!InterlockedFetchAndSetBoolean(&Listener->NeedsDosModeModeEvent)) {
+                QuicListenerReference(Listener);
+                QuicWorkerQueueListener(Listener->Worker, Listener);
```

Comment:
So this is queueing the listener object directly onto the worker? Is this to avoid an allocation and the possibility of OOM failure?